add CVSS feature

therealtoastycat opened the pull request:
Hi - First, thank you for the awesome tool. We have added a few custom features to fit with our internal use and we thought the CVSS v3.0 support might be interesting to others.
In summary:
- Two new fields were added to findings model (CVSS score and CVSS vector)
- UI updated with CVSS calculator, severity automatically updated based on CVSS score
- CVSS score is displayed in report findings list as well as in findings library
- Report writer modified to support the new fields , template.docx was edited to reflect changes

I haven't pushed new severity colors in fixtures as not everyone agrees with CVSS's.

Thank you,

chrismaddalena commented:
@therealtoastycat Thanks for taking the time to submit a PR!

I haven't tested it, but I reviewed the code changes and things seem to be in order. The JS additions are the biggest changes. I know FIRST has a calculator like this on their website. Did you pull the JS from there or do they distribute it?

There are a few things we'll need to consider:

1. We'll need to generate a migration file for the new database fields, but that can wait so migration dependencies and orders don't get disorganized.
2. Rather than in the generic template, I'd prefer the CVSS JS be in its own JS file to be imported as needed.
3. Unit tests need to be updated for the affected models and forms.

Having not seen it in action, I'm not sure if I'm following everything correctly. Is the JS setting the `cvss_score` value for the finding? i.e., you edit a finding, use the calculator, and it creates the score for that field?

We may not want to rely on just JS, if so. We'll be releasing the API soon and someone could edit a finding with the API – i.e., interacting directly with the database. There might need to be a way for it to be calculated server-side.

For anyone looking at this, there is an issue discussing the feature: #153 

I'd like to hear from people who try this or want this to see if it meets their needs. I haven't used anything like CVSS or DREAD scores in a while, so I'm not a good judge for what would make this feature complete.

chrismaddalena commented on ghostwriter/reporting/templates/reporting/report_detail.html:
```diff
@@ -144,6 +145,21 @@ <h4>Current Findings</h4>
                                         title="Click-n-drag to reposition or categorize"
                                     ></td>
                                     <td class="align-middle"><a id="delete-target-content-finding-{{ finding.id }}" class="clickable" href="{% url 'reporting:local_edit' finding.id %}">{{ finding.title }}</a></td>
+                                    <td class="neutral align-middle" style="color:
```
Colors for severity titles should be defined in the CSS files. That makes them consistent across all pages and more easily customizable. A lot of colors like this have been declared here:

https://github.com/GhostManager/Ghostwriter/blob/master/ghostwriter/static/css/base_styles.css#L82

chrismaddalena commented:
Still looking for feedback on this and answers to some of the above items.

This is high on our priority for 2022 following our road map survey, so I'm looking to implement something in the near future.

therealtoastycat commented:
Hi @chrismaddalena - Sorry for the late reply.

I have pushed changes that should resolve 1) and 2), I'll still need to work on 3) next week.

> I know FIRST has a calculator like this on their website. Did you pull the JS from there or do they distribute it?

The files `cvsscal30.js` and `cvsscal30_helptext.js` are FIRST's but their use is permitted as long as we keep the copyright notice in the code.
Everything in file `cvsscalc30_ui.js` is custom.

> Having not seen it in action, I'm not sure if I'm following everything correctly. Is the JS setting the cvss_score value for the finding? i.e., you edit a finding, use the calculator, and it creates the score for that field?
We may not want to rely on just JS, if so. We'll be releasing the API soon and someone could edit a finding with the API – i.e., interacting directly with the database. There might need to be a way for it to be calculated server-side.

Our JS is used to handle integration between the CVSS calculator and the GW user interface. For instance, it automatically modifies the CVSS vector and score fields when you change metrics values, it also changes the finding severity based on the CVSS score.
If someone wants to add a finding with a CVSS score through the API, he would need to know at least the CVSS vector. There is no way to calculate the CVSS score from the vector, but I agree that would require additional server-side code.

Cheers

chrismaddalena commented:
@therealtoastycat Great stuff! Thank you for updating the PR. I'll check it out next week and test the code. I can help get the code ready to merge, too. I haven't had a chance to dig into the CVSS feature just yet. It's next on my list once I have v2.3.0-rc1 out (should be today). I had to get that done so I could record the demos for Black Hat Asia before the April 22 deadline. I'll still be busy with a few things like that this month, but want to merge this feature in May, if possible.

A few notes to help track some items for later:

* We'll need to make sure CVSS data is added to the report data serializer for templates
* Documentation may need updating if Jinja2 statements or expressions change
* Sorting of findings may need adjustment if automatic sorting by score values is desirable

therealtoastycat commented:
Hey @chrismaddalena,

> Unit tests need to be updated for the affected models and forms.

I have updated the factory so that CVSS fields are now included in unit tests for forms and models. Though, I have not added any new tests specific to the CVSS.

> We'll need to make sure CVSS data is added to the report data serializer for templates

The file `reportwriter.py` was updated to enable rendering the new fields in the docx but I didn't take a look to what should be added for pptx.

> Sorting of findings may need adjustment if automatic sorting by score values is desirable

Yes indeed. So far we have been using the "drag & drop" feature to manually sort the findings prior to generate the report :)
If we want to implement this, then maybe the CVSS should have his own model and use weights like it is for the severity.

Cheers,

chrismaddalena commented:
Great stuff, @therealtoastycat! I wasn't able to look at anything last week, but I just finished a few things for the latest release candidate and am now in a spot where I can shift focus for a bit. I plan to review and try out the code changes this week or next.

chrismaddalena commented:
I've got your forked branch\PR checked out and have started reviewing it. I looked over the diffs and they looked pretty good. The above workflow failed because of the version issues between `bleach` and `django-bleach`. I merged the fix for that today with v2.3.0-rc2. There's one conflict, but it's insignificant and easy resolved.

I'd love to get this merged for v2.3.0's final release, so I have been sort of working on this from that angle. Unless someone identifies a problem, v2.3.0-rc2 should be solid. I just need to make some adjustments to the GraphQL permissions. I'll also need to update the GraphQL schema to include the new CVSS fields for findings.

I want to get v2.3.0 released by June, so I promise I'll have feedback for you ASAP, @therealtoastycat :)

chrismaddalena commented:
I've got your forked branch\PR checked out and have started reviewing it. I looked over the diffs and they looked pretty good. The above workflow failed because of the version issues between `bleach` and `django-bleach`. I merged the fix for that today with v2.3.0-rc2. There's one conflict, but it's insignificant and easy resolved.

I'd love to get this merged for v2.3.0's final release, so I have been sort of working on this from that angle. Unless someone identifies a problem, v2.3.0-rc2 should be solid. I just need to make some adjustments to the GraphQL permissions. I'll also need to update the GraphQL schema to include the new CVSS fields for findings.

I want to get v2.3.0 released by June, so I promise I'll have feedback for you ASAP, @therealtoastycat :)

chrismaddalena commented:
@therealtoastycat This is looking really great. I made some adjustments, but nothing big. I moved the calculator inside a collapsible accordion so it is hidden when it isn't needed.

I saw you added reporting `RichText` objects for the CVSS fields (`cvss_score_rt` and `cvss_vector_rt`), but there's no formatting on them. I changed them to work like `severity_rt` and appear with the set severity color value. This way you could write something like `{{ severity_rt }} {{ cvss_score_rt }}` to get `Medium 5.3` all in the same color.

I have just one question for us to consider:

The `cvss_score` field is currently a `CharField` limited to four characters. Perhaps it should be a `FloatField` to only accept float values, e.g., _5.3_.

Thoughts?

Otherwise, we just need to get this working up with the latest code. There's a new `finding_detail.html` template that'll need a section for CVSS info. I can do that after we merge the code, which is also when I need to update the GraphQL schema.

therealtoastycat commented:
Hi @chrismaddalena,

Sorry for the delay, I'm just back from a few weeks off.
I've tried out the new code and the adjustements you made, it looks great! I like the collapsible accordion!

I've pushed a few changes. I deleted the migration file that I previously committed as it was messing up with a newer migration (0027_auto_20220510_1923). Also converted the cvss_score to a FloatField as it makes totally sense to be a float and I updated the `findind_detail.html` file to display the cvss_score.

Let me know if there is anything else I can do.

chrismaddalena commented:
That's great, @therealtoastycat, and no worries about the delay. I finished a few things for v2.3.0 just this week, so the last thing to add to the final release is this PR. I'll work on merging next week. It doesn't look like there will be any conflicts, so it should be pretty easy to merge and then make a few updates.

chrismaddalena commented:
@therealtoastycat Everything is merged and live as part of the v3.0.0-rc1! Thanks for all your work!
